In [1]:
import pandas as pd
import env
import seaborn as sns
import acquire
import prepare
import numpy as np
import matplotlib.pyplot as plt
import scipy
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder, OneHotEncoder, MinMaxScaler
import warnings
warnings.filterwarnings("ignore")

Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | actual cat | actual dog |
|:------------  |-----------:|-----------:|
| predicted cat |         34 |          7 |
| predicted dog |         13 |         46 |

- In the contextof this problem, what is a false positive?

- In the context of this problem, what is a false negative?

- TP: predicted cat, actual is cat (34)
- FP: predicted cat, actual is *__not cat__* (7)
- FN: predicted *__not cat__*, actual is cat (13)
- TN: predicted  *__not cat__*, actual is  *__not cat__* (46)

## Metrics 
- __accuracy__: how well the model does overall
> __(TP + TN) / (TP + TN + FP + FN)__
(34 + 46) / (34 + 7 + 13 +46) = 80.0%
- __precision__: % of actually positive cases that were predicted as positive
> __TP / (TP + FP)__
34 / (34 + 7) = 82.9%
- __recall__: % of of positive predictions that are correct
> __TP / (TP + FN)__
34 / (34 + 13) = 72.2%

- How would you describe this model?
> The model does a good job of predicting hits

#### Exercise 2
 You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

- Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

In [13]:
# Read in csv
df = pd.read_csv('c3.csv')

In [14]:
# Peak at the data
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
actual    200 non-null object
model1    200 non-null object
model2    200 non-null object
model3    200 non-null object
dtypes: object(4)
memory usage: 6.4+ KB


In [16]:
df.nunique()

actual    2
model1    2
model2    2
model3    2
dtype: int64

- I know that:
> - The data has no null values
> - The data types are all as expected
> - There are 200 observations
> - There are only 2 possible outcomes: __Defect and No defect__
- Awesome! Let's add in a baseline to compare our models to. I'll start by figuring out which observation occurs the most

In [17]:
df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [24]:
# add a baseline predictions columns to the df
df['baseline'] = 'No Defect'
# Verify that all the observations in the new column are 'No Defect'
df.baseline.value_counts()

No Defect    200
Name: baseline, dtype: int64

In [25]:
df.head()

actual     model1  model2     model3   baseline
0  No Defect  No Defect  Defect  No Defect  No Defect
1  No Defect  No Defect  Defect     Defect  No Defect
2  No Defect  No Defect  Defect  No Defect  No Defect
3  No Defect     Defect  Defect     Defect  No Defect
4  No Defect  No Defect  Defect  No Defect  No Defect

In [31]:
# Determine accuracy, precision and recall for model1
positive = 'Defect'

# accuracy -- overall hit rate
model_accuracy = (df.model1 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model1 == positive]
model_precision = (subset.model1 == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model1 == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')
print()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')


   model accuracy: 95.00%
baseline accuracy: 92.00%

model precision: 80.00%
baseline precision: nan%

   model recall: 50.00%
baseline recall: 0.00%


In [33]:
# Determine accuracy, precision and recall for model1
positive = 'Defect'

# accuracy -- overall hit rate
model_accuracy = (df.model2 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model2 == positive]
model_precision = (subset.model2 == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model2 == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')
print()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model accuracy: 56.00%
baseline accuracy: 92.00%

model precision: 10.00%
baseline precision: nan%

   model recall: 56.25%
baseline recall: 0.00%


In [34]:
# Determine accuracy, precision and recall for model1
positive = 'Defect'

# accuracy -- overall hit rate
model_accuracy = (df.model3 == df.actual).mean()
baseline_accuracy = (df.baseline == df.actual).mean()

# precision -- how good are our positive predictions?
# precision -- model performance | predicted positive
subset = df[df.model3 == positive]
model_precision = (subset.model3 == subset.actual).mean()
subset = df[df.baseline == positive]
baseline_precision = (subset.baseline == subset.actual).mean()

# recall -- how good are we at detecting actual positives?
# recall -- model performance | actual positive
subset = df[df.actual == positive]
model_recall = (subset.model3 == subset.actual).mean()
baseline_recall = (subset.baseline == subset.actual).mean()


print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')
print()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')
print()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model accuracy: 55.50%
baseline accuracy: 92.00%

model precision: 13.13%
baseline precision: nan%

   model recall: 81.25%
baseline recall: 0.00%


- Use the predictions dataset and pandas to help answer the following questions:

> An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
- positive is __defective__
>> - TP: predicted defective, actual is deffecitve
>> - FP: predicted defective, actual is *__not deffective__*
>> - FN: predicted *__not deffective__*, actual is deffective
>> - TN: predicted  *__not deffectice__*, actual is  *__not deffective__*
- If the goal is to catch as many defective ducks as possible, the cost of a false positive (we predict it's defective but it"s actually not deffective) is not very high.
- The cost of a false negative (we predict it is not deffective when it really is) is high.
- *__Recall__* would be the best metric as it will optimize a reduction in false negatives.
- The best model to use is model 1 as it has the highest rate of precision among all the models and baseline at: 80.0%

> Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
- If the emphasis is on not giving out expensive vacations, the cost of a false positive (we predict it's deffective but it's actually not deffective) is high.
- *__Precision__* would be the best metric as it will optimize a reduction in false positives.
- The best model to use is model 2 as it has the highest recall amonh all the models and baseline at: 81.25%